File: 0951-flip-equivalent-binary-trees/0951-flip-equivalent-binary-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var flipEquiv = function(root1, root2) {
    const flipToEquivalentTree = (node1, node2) => {
        if (!node1 && !node2) return true;
        if (!node1 || !node2) return false;
        if (node1.val !== node2.val) return false;

        const { left, right } = node1;

        if (left?.val !== node2.left?.val) {
            node1.right = left;
            node1.left = right;
        }
        const isLeftEquivalent = flipToEquivalentTree(node1.left, node2.left);
        const isRightEquivalent = flipToEquivalentTree(node1.right, node2.right);

        return isLeftEquivalent && isRightEquivalent;
    };

    return flipToEquivalentTree(root1, root2);
};

File: 0951-flip-equivalent-binary-trees/0951-flip-equivalent-binary-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flipEquiv(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        
        if root1 is None and root2 is None:
            return True

        elif root1 is None or root2 is None:
            return False

        elif root1.val != root2.val:
            return False

        if (root1.left and root2.left and root1.left.val == root2.left.val) or (root1.right and root2.right and root1.right.val == root2.right.val):
            # no swap
            if self.flipEquiv(root1.left, root2.left) and self.flipEquiv(root1.right, root2.right):
                return True

        else:
            # swap
            if self.flipEquiv(root1.left, root2.right) and self.flipEquiv(root1.right, root2.left):
                return True
        
        return False
